### PR TITLE
Java 11 native HTTP client library

### LINKED ISSUE #3433: [REQ] [JAVA] Java 11 Native HTTP client library
##### Description

Support Java 11 Native HTTP client library.

New in the standard library of Java 11 is an [HTTPClient](https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpClient.html) class that simplifies much of the low-level details of making HTTP calls. Using this native client as opposed to a third party client such as Jersey or Feign comes with a few key benefits:

- Reduced set of dependencies for the caller
- Strong commitment to continuous support and backwards compatibility by the Java maintainers
- Up-to-date features such as HTTP2 support

##### openapi-generator version

4.1.0

##### OpenAPI declaration file content or url

Petstore

##### Command line used for generation

Standard

##### Steps to reproduce

There is no Java library that uses the native HTTPClient introduced in Java 11.

##### Related issues/PRs

N/A

##### Suggest a fix/enhancement

We at Yelp have developed a set of mustache templates for use with openapi-generator that produces native Java 11 HTTPClient-based client libraries. We actively use these clients in our production microservices infrastructure and find them robust and performant. We would like to propose integrating these with the main openapi-generator project to open up their use to the wider community.

### PR DESCRIPTION
**Pull request by bbdouglas**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

This is a set of mustache templates to create a new Java library called "native" that is backed by the standard Java 11 [HTTPClient](https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpClient.html) class.

The majority of files are only lightly edited from the default Java templates. The model classes, for example, are entirely untouched. The main two edits are:

- APIClient.mustache
- api.mustache

It follows the pattern used by Jersey, where the APIClient acts as a mechanism for configuring the clients along with a small repository of utility functions. The individual api classes are created from the settings and helper classes held by the APIClient, and are immutable and thread-safe. Object serialization/deserialization is handled by Jackson, as is common in the other libraries.

We at Yelp currently use these templates to generate clientlibs used in our production microservices architecture. This library lacks some of the features seen in other openapi-generator libraries, but represents a fully functional subset for our use case. Additional features could be added on as needed. Here is a summary of the limitations of the current clientlibs:

- No Validation
- No Auth
- No HTML Forms
- Only native Java dates
- Only application/json mimetype

resolves #3433

### CODE REVIEW

**Comment by bbdouglas** [comment]
This is intended to resolve #3433

File: bin/java-petstore-native.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#!/bin/sh
+
+SCRIPT="$0"
+echo "# START SCRIPT: $SCRIPT"
```
Comment (macjohnny) [inline bin/java-petstore-native.sh]: please also add a windows script under `bin/windows/`

Comment (bbdouglas) [reply bin/java-petstore-native.sh]: Done. Thanks for catching that!

**Comment by macjohnny** [comment]
@bbdouglas can you please run `bin/utils/export_generators_readme.sh` and commit the updated docs file `docs/generators/java.md`?
moreover, please also add `samples/client/petstore/java/native/gradlew.bat`

**Comment by bbdouglas** [comment]
@macjohnny I went ahead and completed the tasks you suggested:

- Added an integration test for ApiClient
- Added back gradlew.bat, but without Windows newlines this time (which Travis had complained about before)
- Updated `docs/generators/java.md`

Note, I ran `bin/utils/export_generators_readme.sh`, but it didn't have any effect on the repo. I believe that script changes `docs/generators.md`, but seeing as I'm only adding a library to the existing `java` generator, it didn't have any effect on that file.

**Comment by macjohnny** [comment]
@bbdouglas thanks for the changes!
one last thing: in order to have your tests run in CI, please add them to
https://github.com/OpenAPITools/openapi-generator/blob/2f479f493f47789e8cdcfb83bdf2858271148a38/pom.xml#L1108
and add a `pom.xml` that references your tests, see e.g.
https://github.com/OpenAPITools/openapi-generator/blob/master/samples/client/petstore/java/rest-assured/pom.xml

**Comment by macjohnny** [comment]
and keep in mind https://github.com/OpenAPITools/openapi-generator/pull/3455/files#diff-472c7904fbcc3bf22579387a2039708bR163-R168

**Comment by bbdouglas** [comment]
Hi @macjohnny , thanks for the heads up about #3455 . I went ahead and rebased and added a JSR305 dependency to the pom/gradle templates to support the new `@Nullable` annotation.

I have a pom in `samples/client/petstore/java/native` that runs the tests, but I'm not sure where to reference it in the top-level pom because it needs Java 11 to run. I see profiles for Java 7, 8 and 9, but nothing for Java 11. It looks like we use a combination of Circle CI and Shippable CI, but I don't have much experience with either to know how to configure them.

**Comment by wing328** [comment]
> I have a pom in samples/client/petstore/java/native that runs the tests, but I'm not sure where to reference it in the top-level pom because it needs Java 11 to run. I see profiles for Java 7, 8 and 9, but nothing for Java 11. It looks like we use a combination of Circle CI and Shippable CI, but I don't have much experience with either to know how to configure them.

Very good point. We may need to test it in Shippable or another CI running JDK11+.

I'm in the process to setup another CI (free) so maybe we can use that one to test this new client. More on this later.

**Comment by wing328** [comment]
FYI. Setup drone.io (CI) to cover the Java native client: https://github.com/OpenAPITools/openapi-generator/pull/3472

**Comment by wing328** [comment]
Added @bbdouglas as the template creator via https://github.com/OpenAPITools/openapi-generator/commit/b9e863a9ca538a980a385cb4e1af93cdb4872a6f

**Comment by wing328** [comment]
@bbdouglas thanks for the PR, which has been included in the 4.1.0 release: https://twitter.com/oas_generator/status/1160000504455319553

**Pull request merged**